src/tcp_server.rs

```diff
@@ -1,49 +1,100 @@
+use crate::Kanata;
 use anyhow::Result;
+use net2::TcpStreamExt;
 use parking_lot::Mutex;
-use serde::Serialize;
+use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
+use std::io::Read;
 use std::net::{TcpListener, TcpStream};
+use std::str::FromStr;
 use std::sync::Arc;
+use std::time::Duration;
 
-#[derive(Debug, Serialize)]
-pub enum EventNotification {
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ServerMessage {
     LayerChange { new: String },
 }
 
-impl EventNotification {
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ClientMessage {
+    ChangeLayer { new: String },
+}
+
+impl ServerMessage {
     pub fn as_bytes(&self) -> Result<Vec<u8>> {
         Ok(serde_json::to_string(self)?.as_bytes().to_vec())
     }
 }
 
-pub struct NotificationServer {
+impl FromStr for ClientMessage {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        serde_json::from_str(s)
+    }
+}
+
+pub struct TcpServer {
     pub port: i32,
     pub connections: Arc<Mutex<HashMap<String, TcpStream>>>,
 }
 
-impl NotificationServer {
+impl TcpServer {
     pub fn new(port: i32) -> Self {
         Self {
             port,
             connections: Arc::new(Mutex::new(HashMap::new())),
         }
     }
 
-    pub fn start(&mut self) {
+    pub fn start(&mut self, kanata: Arc<Mutex<Kanata>>) {
         let listener = TcpListener::bind(format!("0.0.0.0:{}", self.port))
             .expect("could not start the tcp server");
 
-        let cl = self.connections.clone();
+        let connections = self.connections.clone();
+
         std::thread::spawn(move || {
             for stream in listener.incoming() {
                 match stream {
-                    Ok(stream) => {
+                    Ok(mut stream) => {
+                        stream
+                            .set_keepalive(Some(Duration::from_secs(30)))
+                            .expect("could not set tcp connection keepalive");
+
                         let addr = stream
                             .peer_addr()
                             .expect("could not find peer address")
                             .to_string();
 
-                        cl.lock().insert(addr, stream);
+                        {
+                            connections.lock().insert(
+                                addr.clone(),
+                                stream.try_clone().expect("could not clone stream"),
+                            );
+                        }
+
+                        log::info!("listening for incoming messages {}", &addr);
+                        loop {
```

Comment:
I believe looping here means that kanata will only ever accept one TCP connection at a time. Since the number of incoming connections should remain small or 0, it should be fine to just spawn a new thread to handle this loop for every connection.

Reply:
This change has been implemented